python-package/lightgbm/sklearn.py

```diff
@@ -1169,6 +1217,19 @@ def feature_names_in_(self) -> np.ndarray:
 class LGBMRegressor(_LGBMRegressorBase, LGBMModel):
     """LightGBM regressor."""
 
+    def _more_tags(self) -> Dict[str, Any]:
+        # handle the case where ClassifierMixin possibly provides _more_tags()
+        if callable(getattr(_LGBMClassifierBase, "_more_tags", None)):
+            tags = _LGBMClassifierBase._more_tags(self)
```

Comment:
`_more_tags` shouldn't care about other classes and the MRO, it should only return what it wants to add, so I'm not sure why this complexity here is needed.

Also, interesting that Classifier tags are needed in the Regressor class

Reply:
> _more_tags shouldn't care about other classes and the MRO, i

This is explained here: https://github.com/microsoft/LightGBM/pull/6651#discussion_r1772666260

I'm not confidence that `RegressorMixin` / `ClassifierMixin` won't add a `_more_tags()`, and I don't want those to silently override LightGBM's preferred tags because we follow `scikit-learn`'s advice to put mixins first in the MRO.

> Also, interesting that Classifier tags are needed in the Regressor class

Thank you! This was a copy-paste mistake. Fixed in https://github.com/microsoft/LightGBM/pull/6651/commits/d1915c0f008a62936ba0aa8f9f66bdd050dc14ec.

It wasn't caught by tests because the tags for `LGBMClassifier`, `LGBMRegressor`, and `LGBMRanker` happen to be the same today.

Reply:
It doesn't matter if they do, your mixin should be the last one in the MRO, and therefore you should rely on what it returns.

Reply:
> and therefore you should rely on what it returns

I'm unsure what you mean by this statement and https://github.com/microsoft/LightGBM/pull/6651#discussion_r1776986827. I'm not sure what "your mixin" means ...is that `LGBMModel` or `sklearn.base.RegressorMixin`?

Let me reduce this to a concrete example, and please tell me what you'd do differently.

```python
class LGBMModel:
    def __sklearn_tags__(self):
        print("--- LGBMModel")

class RegressorMixin:
    def __sklearn_tags__(self):
        print("--- RegressorMixin")
```

We really want to get the LightGBM-specific details from `LGBMModel`. But having `RegressorMixin` "left-most" in MRO as the `scikit-learn` docs suggest means that just using `super()` or not defining the method on `LGBMRegressor` would result in `RegressorMixin`'s definition being used instead if `RegressorMixin` defines `__sklearn_tags__()`.

```python
class LGBMRegressor(RegressorMixin, LGBMModel):
    pass

LGBMRegressor().__sklearn_tags__()
# --- RegressorMixin

class LGBMRegressor(RegressorMixin, LGBMModel):
    def __sklearn_tags__(self):
        super().__sklearn_tags__()

LGBMRegressor().__sklearn_tags__()
# --- RegressorMixin
```

That is why I'm proposing explicitly bypassing the MRO and calling the method from `LGBMModel`.

```python
class LGBMRegressor(RegressorMixin, LGBMModel):
    def __sklearn_tags__(self):
        LGBMModel.__sklearn_tags__(self)

LGBMRegressor().__sklearn_tags__()
# --- LGBMModel
```

What are you suggesting we do differently?